toolchain/driver/compile_subcommand.cpp

```diff
@@ -109,6 +112,25 @@ object output can be forced by enabling `--force-obj-output`.
       },
       [&](auto& arg_b) { arg_b.Set(&output_filename); });
 
+  b.AddOneOfOption(
+      {
+          .name = "opt",
+          .help = R"""(
+Selects the amount of optimization to perform.
+)""",
+      },
+      [&](auto& arg_b) {
+        arg_b.SetOneOf(
+            {
+                // We intentionally don't expose O2 and Os.
```

Comment:
FWIW, as a data point, Chromium uses `-Os` over `-Oz` for most of its components on most build targets (and `-O2` for some components objects that are more perf sensitive): https://source.chromium.org/chromium/chromium/src/+/main:build/config/compiler/BUILD.gn;l=2806;drc=b940e533f03fffbcd16aad86131dcb91abaa21c4

I think we will find that we need to expose both for existing projects to migrate to using Carbon as their C++ toolchain. Maybe as a more advanced option, like `--opt=llvm:Os` which lets you dig down into the specific levels of LLVM breaking past any encapsulation Carbon might provide? Cuz these are presented nicely as you've done here otherwise. Though I wonder what that means for optimizations we write in the Carbon frontend one day.

I am not as sure about `-O2` vs `-O3`. I thought `-O3` is generally avoided for being buggy so I was surprised to see it here and assume my information is outdated. But I see Chromium uses `-O2` for perf sensitive components: https://source.chromium.org/chromium/chromium/src/+/main:build/config/compiler/BUILD.gn;l=2823;drc=b940e533f03fffbcd16aad86131dcb91abaa21c4

Reply:
The suggestion to only support the equivalent of `-Oz` and `-O3`, and not `-Os` nor `-O2` came from @chandlerc; I could only give second-hand rationale for this decision. (I'm personally fine with having both, assuming we can provide reasonable semantic descriptions of what they do, rather than "try `-O2` and `-O3` and pick whichever one is faster in practice" which I've found to be the case for Clang.)

Reply:
I think there are two things we're conflating here...

C++ code built with the packaged Clang toolchain should be able to use the Clang optimization levels exactly. And even using `carbon clang++ -- -Os` and such should work, and we should make sure that *overrides* whatever we set here, at most this should establish a default for a clang invocation.

This flag is more load bearing for Carbon-specific optimization. My hope is that things like generic coalescing and other size improvements will be enough of a size win that in Carbon code we can largely use a single "aggressive" `-O3`.

I'd really like to let users provide a profile, even a file-granularity profile, to control between the finer granularities of -Os, -O2, and -O3 long term. We could even have annotations in the file itself to help control this, rather than having to configure a command line flag differently. Does that make sense?

Reply:
However, it might be nice to leave some documentation to this effect here:

```suggestion
                // We intentionally don't expose O2 and Os. The difference
                // between these levels tends to reflect what achieves the
                // best speed for a specific application, as they all
                // largely optimize for speed as the primary factor.
                //
                // Instead of controlling this with more nuanced flags, we
                // plan to support profile and in-source hints to the
                // optimizer to adjust its strategy in the specific places
                // where the default doesn't have the desired results.
```

Could also consider moving some or all of this to the documentation string instead of a comment.

Reply:
Yes that makes sense and addresses my concerns.